crates/boundless-market/src/contracts/boundless_market.rs

```diff
@@ -1381,6 +1392,133 @@ pub fn decode_calldata(data: &Bytes) -> Result<(Vec<Fulfillment>, AssessorReceip
     ))
 }
 
+#[derive(Debug, Clone)]
+/// Represents the parameters for submitting a Merkle Root.
+pub struct Root {
+    /// The address of the set verifier contract.
+    pub verifier_address: Address,
+    /// The Merkle root of the proof.
+    pub root: B256,
+    /// The seal of the proof.
+    pub seal: Bytes,
+}
+
+#[derive(Debug, Clone)]
+/// Represents the parameters for pricing a request.
+pub struct Price {
+    /// The list of requests to be priced.
+    pub requests: Vec<ProofRequest>,
+    /// The list of client signatures for the requests.
+    pub client_sigs: Vec<Bytes>,
+}
+
+#[derive(Clone)]
+#[non_exhaustive]
+/// Builder for creating a fulfillment request.
+pub struct FulfillmentBuilder<P> {
+    /// The Boundless market service instance.
+    pub boundless_market: BoundlessMarketService<P>,
+    /// The parameters for submitting a Merkle Root
+    pub root: Option<Root>,
+    /// The parameters for pricing a request
+    pub price: Option<Price>,
+    /// The fulfillments to be submitted
+    pub fulfillments: Vec<Fulfillment>,
+    /// The assessor receipt
+    pub assessor_receipt: AssessorReceipt,
+    /// Whether to withdraw the fee
+    pub withdraw: bool,
+}
+
+impl<P: Provider> FulfillmentBuilder<P> {
+    /// Creates a new instance of the `Fulfill` struct.
+    pub fn new(
+        boundless_market: BoundlessMarketService<P>,
+        fulfillments: Vec<Fulfillment>,
+        assessor_receipt: AssessorReceipt,
+    ) -> Self {
+        Self {
+            boundless_market,
+            root: None,
+            price: None,
+            fulfillments,
+            assessor_receipt,
+            withdraw: false,
+        }
+    }
+
+    /// Sets the parameters for submitting a Merkle Root.
+    pub fn with_submit_root(self, verifier_address: Address, root: B256, seal: Bytes) -> Self {
+        Self { root: Some(Root { verifier_address, root, seal }), ..self }
+    }
+
+    /// Sets the parameters for pricing a request.
+    pub fn with_price(self, requests: Vec<ProofRequest>, client_sigs: Vec<Bytes>) -> Self {
+        Self { price: Some(Price { requests, client_sigs }), ..self }
+    }
```

Comment:
Based on my comment above, I would recommend we call this `with_unlocked_request`.